packages/syft/src/syft/core/tensor/smpc/share_tensor.py

```diff
@@ -0,0 +1,156 @@
+
+
+@bind_protobuf
+class ShareTensor(PassthroughTensor, Serializable):
+    def __init__(
+        self,
+        rank: int,
+        ring_size: int = 2 ** 64,
+        value: Optional[Any] = None,
+    ) -> None:
+        self.rank = rank
+        self.ring_size = ring_size
+        self.min_value, self.max_value = ShareTensor.compute_min_max_from_ring(
+            self.ring_size
+        )
+        super().__init__(value)
+
+    def copy_tensor(self):
+        return ShareTensor(rank=self.rank, ring_size=self.ring_size)
+
+    @staticmethod
+    @lru_cache(32)
+    def compute_min_max_from_ring(ring_size: int = 2 ** 64) -> Tuple[int, int]:
+        min_value = (-ring_size) // 2
+        max_value = (ring_size - 1) // 2
+        return min_value, max_value
+
+    """ TODO: Remove this -- we would use generate_przs since the scenario we are testing is that
+    the secret is remotly
+    @staticmethod
+    def generate_shares(secret, nr_shares, ring_size=2 ** 64):
+        from .fixed_precision_tensor import FixedPrecisionTensor
+
+        if not isinstance(secret, (int, FixedPrecisionTensor)):
+            secret = FixedPrecisionTensor(value=secret)
+
+        shape = secret.shape
+        min_value, max_value = ShareTensor.compute_min_max_from_ring(ring_size)
+
+        generator_shares = np.random.default_rng()
+
+        random_shares = []
+        for i in range(nr_shares):
+            random_value = generator_shares.integers(
+                low=min_value, high=max_value, size=shape
+            )
+            fpt_value = FixedPrecisionTensor(value=random_value)
+            random_shares.append(fpt_value)
+
+        shares_fpt = []
+        for i in range(nr_shares):
+            if i == 0:
+                share = value = random_shares[i]
+            elif i < nr_shares - 1:
+                share = random_shares[i] - random_shares[i - 1]
+            else:
+                share = secret - random_shares[i - 1]
+
+            shares_fpt.append(share)
+
+        # Add the ShareTensor class between them
+        shares = []
+        for rank, share_fpt in enumerate(shares_fpt):
+            share_fpt.child = ShareTensor(rank=rank, value=share_fpt.child)
+            shares.append(share_fpt)
+
+        return shares
+    """
+
+    @staticmethod
+    def generate_przs(
+        value: Optional[Any],
+        shape: Tuple[int],
+        rank: int,
+        nr_parties: int,
+        seed_shares: int,
+    ) -> "ShareTensor":
+        # syft absolute
+        from syft.core.tensor.tensor import Tensor
+
+        if value is None:
+            value = Tensor(np.zeros(shape, dtype=np.int64))
+
+        # TODO: Sending the seed and having each party generate the shares is not safe
+        # Since the parties would know some of the other parties shares (this might not impose a risk
+        # when shares are not sent between parties -- like private addition/subtraction, but it might
+        # impose for multiplication
```

Comment:
Q: Why is it insecure, since we beaver triples, I think it should be secure.

Reply:
Because the parties might know how the shares from the other parties might look.
Since they all get the same seed and generate shares (even thought they do not keep them -- they could still "look" at them)